### ISSUE TITLE
ATProto => AP: reply to a bridged Bsky comment in a Fedi thread not bridged to Fedi

### ISSUE THREAD

**Issue by mackuba** [snarfed/bridgy-fed#1218]
There's a thread starting from this post:  https://attractive.space/@Setok/112870422629728861

`@nafnlaus.bsky.social` replied, and then there's some back and forth between `@nafnlaus` and OP `@Setok@attractive.space`. So far all replies are seen in both places.

Then, `@nafnlaus` writes this post https://bsky.app/profile/nafnlaus.bsky.social/post/3kygrzepb5x23 ("First user-run appviews (web frontends) were spun out"), which again, is seen in both places.

Finally, I join in and make this reply https://bsky.app/profile/mackuba.eu/post/3kyh6mcfno72q, replying to that last post. If you look at the OP post's thread page at https://attractive.space/@Setok/112870422629728861, this reply isn't available there. Another my reply, replying directly to the OP, did get delivered.

I'm not sure if this is intended behavior, but from my rough understanding of AP, this feels like the classic Mastodon problem of "not all thread replies are visible on each instance"… So here, BF probably has seen that the comment I'm replying to was a bridged one, so didn't deliver it to any AP instance because the direct target wasn't an AP source.

But "normal", "native" replies made by anyone to someone down the thread (say, by `c@cc.social` to `b@bb.social` on a thread started by `a@aa.social`) are delivered not only to the direct parent (`bb.social`), but also to the OP (`aa.social`) and also to anyone folllowing `a@aa.social`, who will see that thread on their instances in the home feed, so I feel like this reply should also be delivered to all those places?…

**Comment by snarfed** [comment]
Detailed sleuthing!

I have to admit, given how unreliable reply delivery is in the "native" fediverse, as you mention, I don't have a ton of interest in debugging it here. Having said that, my guess is that the extra down-thread deliveries you're seeing here are [inbox forwarding](https://www.w3.org/TR/activitypub/#inbox-forwarding), which BF doesn't yet accept or send mostly because it doesn't yet do LD Signatures, [which are difficult to verify](https://github.com/snarfed/bridgy-fed/issues/566#issuecomment-1936399383).

**Comment by qazmlp** [comment]
I don't think that's it, Mastodon just sends the replies directly also to the servers of posters earlier in the conversation.
(I'm not sure if this is restricted to mentions or if it goes beyond that, but I *suspect* it does.)

Mastodon itself [recommends against](https://docs.joinmastodon.org/spec/security/#ld) implementing LD signatures too, so I'd be cautious about trying to fix this issue that way.

---

I'm not sure what the best implementation here is, since iirc Bluesky doesn't have mention control like Mastodon does, but Mastodon expects some level of tag to notify users at all. (You can reply without notifying even the direct poster to whom you're replying, by deleting the @-mention.)

**Comment by snarfed** [comment]
It's worth distinguishing _delivery_ from _user notification_ here, ie in a notifications tab/screen. Afaict this issue is about delivery.

> I think at the very least I'd expect the thread root's instance to be notified.

That would definitely be nice! A couple thoughts:

* As an example, Bluesky makes this easy. Every reply has a direct link to its root post as well as its parent. AP, on the other hand, doesn't. You have to follow the `inReplyTo` chain up, one reply at a time, fetching each one from its own instance, before you eventually find the root, and the length of that chain is unbounded. So, I'm reluctant to do this for every reply.
* We don't necessarily have to make a judgment call here. The applicable bits of standards are [AP inbox delivery](https://www.w3.org/TR/activitypub/#delivery) and [AS2 audience targeting](https://www.w3.org/TR/activitystreams-vocabulary/#audienceTargeting). They're still a bit incomplete - afaik there's nothing normative that says how to populate `to` and `cc` for a reply - but I don't know of any fediverse server that walks the full reply chain to find the root and adds its inbox. Looking at an example Mastodon reply, it has `to: Public` and `cc`'s the parent (me) and the author's followers:
    ````
    "to": ["https://www.w3.org/ns/activitystreams#Public"],
    "cc": [
      "https://social.ayco.io/users/ayo/followers",
      "https://fed.brid.gy/snarfed.org"
    ],
    ````

**Comment by snarfed** [comment]
> I don't think that's it, Mastodon just sends the replies directly also to the servers of posters earlier in the conversation.

@qazmlp interesting! Out of curiosity, do you actually know this? And are you talking about inbox delivery specifically, or where replies are eventually visible? Downstream replies famously aren't always visible on every instance, even instances that were directly in their upstream chain, so it's surprising to hear that Mastodon might do inbox delivery to every upstream post in a thread. (If that's true, it would seem to be broken. 😁)

**Comment by mackuba** [comment]
> It's worth distinguishing _delivery_ from _user notification_ here, ie in a notifications tab/screen. Afaict this issue is about delivery.

Yeah yeah, I meant it should be sent there to the instance so it's visible there :) I'm not very familiar with the rules of how this delivery of comments works, but I'd think that the OP who started the thread should be able to see the complete tree of all comments below on their instance, even not those made directly to them.

**Comment by mackuba** [comment]
> * Bluesky makes this easy: every reply has a direct link to its root post as well as its parent. AP, on the other hand, doesn't. You'd have to follow the `inReplyTo` chain up, one reply at a time, fetching each one from its own instance, before you eventually find the root, and the length of that chain is unbounded. So, I'm reluctant to do this for every reply.

Oof, ok, yeah that sounds like it complicates things...

**Comment by qazmlp** [comment]
> > I don't think that's it, Mastodon just sends the replies directly also to the servers of posters earlier in the conversation.
> 
> @qazmlp interesting! Out of curiosity, do you actually know this? And are you talking about inbox delivery specifically, or where replies are eventually visible? Downstream replies famously aren't always visible on every instance, even instances that were directly in their upstream chain, so it's surprising to hear that Mastodon might do inbox delivery to every upstream post in a thread. (If that's true, it would seem to be broken. 😁)

As I wrote:

> (I'm not sure if this is restricted to mentions or if it goes beyond that, but I suspect it does.)

I did look it up just now, however: https://github.com/mastodon/mastodon/blob/4eab6aef95554d74237fcdbe13c986373a643739/app/lib/status_reach_finder.rb

So no, it does not necessarily deliver to inboxes of earlier users who weren't mentioned, but it is much more aggressive about who gets mentioned than Bridgy Fed appears to be. It prefills mentions to the in-reply-to-author and also everyone who was mentioned in the previous post.

I'm not sure what the best mapping is here - does Bluesky notify users of replies to posts in which they were just mentioned? On Mastodon it would be customary to untag them in many cases - but not all. I wish Bluesky had checkboxes here like the ones Twitter has.

(It *is* possible to mute a conversation in Mastodon too (via the post notification specifically), but that button isn't available in the mobile app (and afaik no third party apps) for whatever reason, so this isn't a reliable capability.)

**Comment by Tamschi** [comment]
Bridgy Fed should probably walk upwards through the conversation and add `"cc"` mentions for:

- each Bsky author it sees **until**
- when it sees the first AP-native post, from which it should use the author and all `"cc"` mentions.

I think that matches Mastodon's behaviour then and allows at least ActivityPub users to untag someone partway through the conversation.

**Comment by snarfed** [comment]
Mastodon does have that effect, but it seems like the way it does it is just by @-mentioning the OP and everyone mentioned in the immediate in-reply-to post, right? Not by actually walking the reply chain and collecting authors?

I definitely could walk the reply chain like this for every reply bridged into the fediverse, but that would be pretty heavyweight, so I'm reluctant.

**Comment by Tamschi** [comment]
Yes, it's a protocol difference. I don't think there's an easier way to achieve parity here… aside from caching this information ahead of time, perhaps, or maybe you could store a thread ID for each post to look up only the parent and then all posts from the entire thread in a second query. Or maybe the database you're using supports something like Recursive Common Table Expressions and you get graph traversal very cheaply after all.

Mastodon just mentions the parent post's author and everyone mentioned there, yes. The problem is that I'm fairly sure Bluesky posts won't normally have @-mentions to previous posters, so you could easily miss the information.

**Comment by snarfed** [comment]
@MetaflameDragon raised a more specific case of this that we could consider: if a fediverse post is bridged into Bluesky, and a Bluesky user replies, and then replies again and threads a reply chain, we only deliver the first reply to the OP's inbox. We should probably detect the self-reply thread and deliver all of them.

**Comment by mackuba** [comment]
Yeah, I think this is pretty important, because people reply to Mastodon posts with subthreads and then don't realize that they don't get fully bridged (e.g.: https://bsky.app/profile/wjmaggos.liberal.city.ap.brid.gy/post/3logbje2a4cd2)

**Comment by snarfed** [comment]
https://github.com/snarfed/bridgy-fed/issues/1608 will help some here, at least for delivering replies to more upstream fediverse instances specifically.

**Comment by snarfed** [comment]
I think #1608 actually fixed this. It does what Mastodon itself does, as described in https://github.com/snarfed/bridgy-fed/issues/1218#issuecomment-2450240287 :

> Mastodon does have that effect, but it seems like the way it does it is just by @-mentioning the OP and everyone mentioned in the immediate in-reply-to post, right? Not by actually walking the reply chain and collecting authors?

**Comment by mackuba** [comment]
Awesome! I'll get back to you if this keeps happening.

**Comment by snarfed** [comment]
Thanks! And it's definitely still possible. The current implementation depends on fediverse servers all adding and/or keeping mentions in replies for all upstream authors. That's pretty standard, but not required, and probably not universal, and users can remove them manually. But we're hopefully at par with that native fediverse behavior now.

**Comment by mackuba** [comment]
So, for example, this one: https://bsky.app/profile/mackuba.eu/post/3lwn46dvd222z didn't get delivered here: https://infosec.exchange/@paulehoffman/115046664068926114 - is this not the case we're talking about here?

**Comment by snarfed** [comment]
Aha, yes! Thanks, good example of a case that probably still needs work, and that #1608 doesn't handle since that fix is ActivityPub-specific.

**Comment by mackuba** [comment]
Oof… wait so why doesn't the ActivityPub fix work here?

**Comment by snarfed** [comment]
Yeah, good question. It only looks at the immediate parent to find extra targets to deliver to, not all the way up the thread. So in this case, your first reply got delivered to infosec.exchange, but your second didn't, since you were replying to yourself.

One possible fix would be for Bridgy Fed to store those extra targets with each reply, so that subsequent replies pick them up.
